subsys/fs/ext2/ext2_bitmap.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include "ext2_bitmap.h"
+
+LOG_MODULE_DECLARE(ext2);
+
+/* NOTICE: Offsets in bitmap start with 0 */
+
+int ext2_bitmap_get(uint8_t *bm, uint32_t index, uint32_t size)
+{
+	uint32_t idx = index / 8;
+	uint32_t off = index % 8;
+
+	if (idx >= size) {
+		LOG_ERR("(get) Tried to access value outside of bitmap (%d)", index);
+		return -EINVAL;
+	}
+
+	uint8_t b = bm[idx];
+
+	return b & BIT(off);
+}
+
+int ext2_bitmap_set(uint8_t *bm, uint32_t index, uint32_t size)
+{
+	LOG_DBG("Setting %d bit in bitmap", index);
+
+	uint32_t idx = index / 8;
+	uint32_t off = index % 8;
+
+	if (idx >= size) {
+		LOG_ERR("Tried to set value outside of bitmap (%d)", index);
+		return -EINVAL;
+	}
+
+	__ASSERT((bm[idx] & BIT(off)) == 0, "Bit %d set in bitmap", index);
+
+	LOG_DBG("Bitmap %d: %x", idx, bm[idx]);
+	bm[idx] |= BIT(off);
+	LOG_DBG("Bitmap %d: %x", idx, bm[idx]);
+
+	return 0;
+}
+
+int ext2_bitmap_unset(uint8_t *bm, uint32_t index, uint32_t size)
```

Comment:
This has similar issues as `ext2_bitmap_set` that is only called by `ext2_free_block`  and `ext2_free_inode`
As far as I can tell the `index` is quite processed before it reaches the function, whatever form it takes in a caller.
The error message so deep is quite useless, because it states nothing, and considering the fact that there are many places the code with bad `index` can topple over before this is even called, is this error message even possible?